projects/cm_mcu/LocalTasks.c

```diff
@@ -612,28 +687,96 @@ TickType_t getFFupdateTick(int mask)
   }
 }
 
-int8_t getFFtemp(const uint8_t i)
+uint16_t getFFtemp(const uint8_t i)
 {
   int i1 = 1;
   int8_t val;
   configASSERT(i < NFIREFLIES);
   if (i < NFIREFLIES_IT_F1) {
     int index = i * (ffl12_f1_args.n_commands * ffl12_f1_args.n_pages) + i1;
-    val = (int8_t)ffl12_f1_args.sm_values[index];
+    val = ffl12_f1_args.sm_values[index];
   }
 
   else if (NFIREFLIES_IT_F1 <= i && i < NFIREFLIES_IT_F1 + NFIREFLIES_DAQ_F1) {
     int index = (i - NFIREFLIES_IT_F1) * (ffldaq_f1_args.n_commands * ffldaq_f1_args.n_pages) + i1;
-    val = (int8_t)ffldaq_f1_args.sm_values[index];
+    val = ffldaq_f1_args.sm_values[index];
   }
 
   else if (NFIREFLIES_F1 <= i && i < NFIREFLIES_F1 + NFIREFLIES_IT_F2) {
     int index = (i - NFIREFLIES_F1) * (ffl12_f2_args.n_commands * ffl12_f2_args.n_pages) + i1;
-    val = (int8_t)ffl12_f2_args.sm_values[index];
+    val = ffl12_f2_args.sm_values[index];
   }
   else {
     int index = (i - NFIREFLIES_F1 - NFIREFLIES_IT_F2) * (ffldaq_f2_args.n_commands * ffldaq_f2_args.n_pages) + i1;
-    val = (int8_t)ffldaq_f2_args.sm_values[index];
+    val = ffldaq_f2_args.sm_values[index];
+  }
+
+  return val;
+}
+
+uint16_t getFFoptpow(const uint8_t i)
+{
+
+  uint16_t avg_val = 0;
+  uint16_t sum_val = 0;
+  configASSERT(i < NFIREFLIES);
+  if (i < NFIREFLIES_IT_F1) {
+    for (int i1 = 4; i1 < ffl12_f1_args.n_commands; ++i1) {
+      int dev = i - ff_moni2c_arg[0].int_idx + ff_moni2c_arg[0].dev_int_idx;
+      int index = dev * (ffl12_f1_args.n_commands * ffl12_f1_args.n_pages) + i1;
+      sum_val += ffl12_f1_args.sm_values[index];
+    }
+    avg_val = sum_val / (ffl12_f1_args.n_commands - 4);
+  }
+
+  else if (NFIREFLIES_IT_F1 <= i && i < NFIREFLIES_IT_F1 + NFIREFLIES_DAQ_F1) {
+    for (int i1 = 4; i1 < ffldaq_f1_args.n_commands; ++i1) {
+      int dev = i - ff_moni2c_arg[1].int_idx + ff_moni2c_arg[1].dev_int_idx;
+      int index = (dev) * (ffldaq_f1_args.n_commands * ffldaq_f1_args.n_pages) + i1;
+      sum_val += ffldaq_f1_args.sm_values[index];
+    }
+    avg_val = sum_val / (ffldaq_f1_args.n_commands - 4);
+  }
+
+  else if (NFIREFLIES_F1 <= i && i < NFIREFLIES_F1 + NFIREFLIES_IT_F2) {
+    for (int i1 = 4; i1 < ffl12_f2_args.n_commands; ++i1) {
+      int dev = i - ff_moni2c_arg[2].int_idx + ff_moni2c_arg[2].dev_int_idx;
+      int index = (dev) * (ffl12_f2_args.n_commands * ffl12_f2_args.n_pages) + i1;
+      sum_val += ffl12_f2_args.sm_values[index];
+    }
+    avg_val = sum_val / (ffl12_f2_args.n_commands - 4);
+  }
+  else {
+    for (int i1 = 4; i1 < ffldaq_f2_args.n_commands; ++i1) {
+      int dev = i - ff_moni2c_arg[3].int_idx + ff_moni2c_arg[3].dev_int_idx;
+      int index = (dev) * (ffldaq_f2_args.n_commands * ffldaq_f2_args.n_pages) + i1;
+      sum_val += ffldaq_f2_args.sm_values[index];
+    }
+    avg_val = sum_val / (ffldaq_f2_args.n_commands - 4);
```

Comment:
a lot of copypasta here. can you clean this up by maybe usinga  pointer for the ff*args in the repeated code, and just setting the appropriate pointer 1x?

Reply:
Done